File: flutter_ui/lib/backend/platform_base.dart
import 'package:flutter/foundation.dart';
import 'dart:io' show Platform;

class PlatformBase {
  static const int KEY_CODE_TAB = 9;
  static const int KEY_CODE_UP = 38;
  static const int KEY_CODE_DOWN = 40;
  static const int KEY_CODE_LEFT = 37;
  static const int KEY_CODE_RIGHT = 39;
  static const int KEY_CODE_SPACE = 32;
  static const int KEY_CODE_ENTER = 13;
  static const int KEY_CODE_ESCAPE = 27;

  PlatformBase();

  bool get isWindows =>
      !kIsWeb && defaultTargetPlatform == TargetPlatform.windows;
  bool get isLinux => !kIsWeb && defaultTargetPlatform == TargetPlatform.linux;
  bool get isMac => !kIsWeb && defaultTargetPlatform == TargetPlatform.macOS;
  bool get isMobile =>
      !kIsWeb &&
      (defaultTargetPlatform == TargetPlatform.android ||
          defaultTargetPlatform == TargetPlatform.iOS);

  double getMetric(String id) => 0;
  int? getColor(String id) => null;
  String? getSystemLocale() => kIsWeb ? null : Platform.localeName;
  double perf() => DateTime.now().millisecondsSinceEpoch.toDouble();

  int get KeyTab => getKeyCode("tab");
  int get KeyUp => getKeyCode("up");
  int get KeyDown => getKeyCode("down");
  int get KeyLeft => getKeyCode("left");
  int get KeyRight => getKeyCode("right");
  int get KeySpace => getKeyCode("space");
  int get KeyEnter => getKeyCode("enter");
  int get KeyEscape => getKeyCode("escape");

  int getKeyCode(String keyId) {
    switch (keyId) {
      case "tab":
        return KEY_CODE_TAB;
      case "up":
        return KEY_CODE_UP;
      case "down":
        return KEY_CODE_DOWN;
      case "left":
        return KEY_CODE_LEFT;
      case "right":
        return KEY_CODE_RIGHT;
      case "space":
        return KEY_CODE_SPACE;
      case "enter":
        return KEY_CODE_ENTER;
      case "escape":
        return KEY_CODE_ESCAPE;
      default:
        return keyId.isNotEmpty ? keyId.codeUnitAt(0) : -1;
    }
  }
}

File: luaui/backend/PlatformBase.lua
local PlatformBase = {}
PlatformBase.__index = PlatformBase
function PlatformBase.new() return setmetatable({ _isMobile = nil }, PlatformBase) end
function PlatformBase:get_isWindows() return love.system.getOS() == "Windows" end
function PlatformBase:get_isLinux() return love.system.getOS() == "Linux" end
function PlatformBase:get_isMac() return love.system.getOS() == "OS X" end
function PlatformBase:get_isMobile()
    if self._isMobile == nil then
        local os = love.system.getOS()
        self._isMobile = (os == "Android" or os == "iOS")
    end
    return self._isMobile
end
function PlatformBase:perf() return love.timer.getTime() * 1000 end
function PlatformBase:getKeyCode(keyId) return string.byte(keyId, 1) end
return PlatformBase
